pkg/agent/controller/agent.go

```diff
@@ -0,0 +1,207 @@
+		servicesSynced:   config.ServiceInformer.Informer().HasSynced,
+
+		excludeNamespaces: exclusionNSMap,
+	}
+
+	klog.Info("Setting up event handlers")
+	config.ServiceInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: agentController.enqueueService,
+		UpdateFunc: func(old, newObj interface{}) {
+			agentController.handleUpdateService(old, newObj)
+		},
+		DeleteFunc: agentController.handleRemovedService,
+	}, 0)
+
+	return agentController
+}
+
+func (a *Controller) Run(stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+
+	// Start the informer factories to begin populating the informer caches
+	klog.Info("Starting Agent controller")
+
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, a.servicesSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	go wait.Until(a.runServiceWorker, time.Second, stopCh)
+	klog.Info("Lighthouse agent workers started")
+	<-stopCh
+	klog.Info("Lighthouse Agent stopping")
+	return nil
+}
+
+func (a *Controller) enqueueService(obj interface{}) {
+	if key := a.getEnqueueKey(obj); key != "" {
+		klog.V(4).Infof("Enqueueing service %v for agent", key)
+		a.serviceWorkqueue.AddRateLimited(key)
+	}
+}
+
+func (a *Controller) getEnqueueKey(obj interface{}) string {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return ""
+	}
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return ""
+	}
+	if !a.isValidService(ns, name) {
+		return ""
+	}
+	return key
+}
+
+func (a *Controller) isValidService(namespace string, serviceName string) bool {
+	if a.excludeNamespaces[namespace] || excludeSvcList[serviceName] {
+		return false
+	}
+	return true
+}
+
+func (a *Controller) runServiceWorker() {
+	for a.processNextService() {
+
+	}
+}
+
+func (a *Controller) processNextService() bool {
+	obj, shutdown := a.serviceWorkqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func() error {
+		defer a.serviceWorkqueue.Done(obj)
+
+		key := obj.(string)
+		ns, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			a.serviceWorkqueue.Forget(obj)
+			return fmt.Errorf("error while splitting meta namespace key %s: %v", key, err)
+		}
+
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of object before attempting update
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			service, err := a.kubeClientSet.CoreV1().Services(ns).Get(name, metav1.GetOptions{})
+
+			if err != nil {
+				if errors.IsNotFound(err) {
+					// already deleted Forget and return
+					return nil
+				} else {
+					// could be API error, so we want to requeue
+					logAndRequeue(key, a.serviceWorkqueue)
```

Comment:
`logAndRequeue` is called on L141 so the else clause here can be removed.

Reply:
Good call. missed this.